### PR TITLE
[Refactor] Slippage Query Handling

### PR DESCRIPTION
**Pull request by bh2smith**
There were a lot of weird things happening

- Moving test only classes and models into the test directory.
- Simplifying e2e tests to remove a bunch of code redundancy
- Unit testing some edge cases that were forgotten
- Renaming a few queries

No logical changes have been made (apart from raising an error when token list is empty).

There are a bunch of inline comments here describing what is happening.

# Test Plan 

CI

### CODE REVIEW

File: src/models.py [hunk 1]
```diff
@@ -76,35 +75,3 @@ def from_str(cls, type_str: str) -> TransferType:
             return cls[type_str.upper()]
         except KeyError as err:
             raise ValueError(f"No TransferType {type_str}!") from err
-
-
-@dataclass
-class InternalTokenTransfer:
```
Comment (bh2smith) [inline src/models.py:82]: This class was only used for e2e testing and has been moved there.

File: tests/e2e/test_internal_trades.py [hunk 2]
```diff
@@ -15,15 +51,23 @@ def token_slippage(
...
+    select_transfers_query = """
+        select block_time,
+           CONCAT('0x', ENCODE(tx_hash, 'hex')) as tx_hash,
+           solver_address,
+           solver_name,
+           CONCAT('0x', ENCODE(token, 'hex')) as token,
+           amount,
+           transfer_type
+        from incoming_and_outgoing_with_buffer_trades
+    """
```
Comment (bh2smith) [inline tests/e2e/test_internal_trades.py:69]: There is no reason to have this in a file since it is only ever used here.

File: src/fetch/period_slippage.py [hunk 3]
```diff
@@ -28,22 +31,43 @@ def prepend_to_sub_query(query: str, table_to_add: str) -> str:
...
+    Default query type input it total, but we can request
+    per transaction results for testing
+    """
+
+    slippage_sub_query = DuneAnalytics.open_query("./queries/period_slippage.sql")
+    select_statement = f"""
+    select *, 
+        usd_value / (select price from eth_price) * 10 ^ 18 as eth_slippage_wei 
+    from {query_type}
+    """.strip()
```
Comment (bh2smith) [inline src/fetch/period_slippage.py:64]: This removes the need to pass the {{ResultTable}} as a query parameter (since we are dynamically using the string right here)

File: tests/unit/test_token_list_table.py [hunk 4]
```diff
@@ -1,28 +1,31 @@
...
 
 class TestQueryBuilding(unittest.TestCase):
+    def test_empty_token_list(self):
+        with self.assertRaises(ValueError) as err:
+            allowed_token_list_query([])
+        self.assertEqual(str(err.exception), "Cannot build query for empty token list")
+
     def test_builds_intended_query(self):
-        list = [
+        tokens = [
```
Comment (bh2smith) [inline tests/unit/test_token_list_table.py:13]: list is a reserved name in python. should never be used as a variable name!

File: tests/e2e/test_internal_trades.py [hunk 5]
```diff
@@ -15,15 +49,23 @@ def token_slippage(
...
-    path = "./queries/slippage"
-    select_transfers_query = dune.open_query(
-        File("select_in_out_with_buffers.sql", path).filename()
-    )
-    slippage_sub_query = dune.open_query(
-        File("subquery_batchwise_internal_transfers.sql", path).filename()
-    )
+    slippage_sub_query = dune.open_query("./queries/period_slippage.sql")
+    select_transfers_query = """
+        select block_time,
```
Comment (josojo) [inline tests/e2e/test_internal_trades.py:59]: To be honest, I dislike it.
having all dune queries in one place makes them very easily searchable and understandable. I think the overhead of file-reading for them is negligble

File: src/fetch/period_slippage.py [hunk 6]
```diff
@@ -3,17 +3,20 @@
...
-from src.file_io import File
 from src.models import Address, Network
-from src.token_list import ALLOWED_TOKEN_LIST_URL, get_trusted_tokens_from_url
+from src.token_list import fetch_trusted_tokens
 from src.utils.script_args import generic_script_init
 
 
 def allowed_token_list_query(token_list: list[str]) -> str:
     """Constructs sub query for allowed tokens"""
+    if len(token_list) == 0:
```
Comment (josojo) [inline src/fetch/period_slippage.py:17]: This is very helpful. thanks

File: tests/e2e/test_internal_trades.py (hunk 5 above)
Comment (bh2smith) [reply tests/e2e/test_internal_trades.py]: Fixed this.

File: src/fetch/period_slippage.py [hunk 7]
```diff
@@ -2,17 +2,20 @@
...
 
 from dataclasses import dataclass
+from enum import Enum
 from pprint import pprint
 
 from src.dune_analytics import DuneAnalytics, QueryParameter
-from src.file_io import File
 from src.models import AccountingPeriod, Address, Network
-from src.token_list import ALLOWED_TOKEN_LIST_URL, get_trusted_tokens_from_url
+from src.token_list import fetch_trusted_tokens
```
Comment (bh2smith) [inline src/fetch/period_slippage.py:10]: The url is a constant and can be entirely contained within the method that uses it.

File: src/fetch/period_slippage.py [hunk 8]
```diff
@@ -27,22 +30,43 @@ def prepend_to_sub_query(query: str, table_to_add: str) -> str:
 
 def add_token_list_table_to_query(original_sub_query: str) -> str:
     """Inserts the token_list table right after the WITH statement into the sql query"""
-    token_list = get_trusted_tokens_from_url(ALLOWED_TOKEN_LIST_URL)
-    sql_query_for_allowed_token_list = allowed_token_list_query(token_list)
-    return prepend_to_sub_query(original_sub_query, sql_query_for_allowed_token_list)
+    token_list = fetch_trusted_tokens()
+    allowed_tokens_query = allowed_token_list_query(token_list)
+    return prepend_to_sub_query(original_sub_query, allowed_tokens_query)
```
Comment (bh2smith) [inline src/fetch/period_slippage.py:35]: Shortening the name of some functions and variables that were too long.

File: tests/e2e/test_internal_trades.py [hunk 9]
```diff
@@ -1,66 +1,113 @@
...
     token_str: str,
-    internal_trade_list: list[InternalTokenTransfer],
+    internal_trade_list: list[InternalTransfer],
 ) -> int:
     return sum(a.amount for a in internal_trade_list if a.token == Address(token_str))
 
 
-def get_internal_transfers(
-    dune: DuneAnalytics, tx_hash: str, period_start: datetime, period_end: datetime
-) -> list[InternalTokenTransfer]:
```
Comment (bh2smith) [inline tests/e2e/test_internal_trades.py:19]: This method was moved inside the test class which has most of these input parameters are attributes. Now we can call this method as `self.get_internal_transfers` passing in only the `tx_hash`

**Review by MartinquaXD — APPROVED** [review]
Overall it makes sense to declutter the API but I think this probably could have been multiple separate PRs.

File: src/fetch/period_slippage.py [hunk 10]
```diff
@@ -27,22 +30,43 @@ def prepend_to_sub_query(query: str, table_to_add: str) -> str:
...
+    """
+
+    PER_TX = "results_per_tx"
+    TOTAL = "results"
+
+    def __str__(self) -> str:
+        return self.value
+
+
+def slippage_query(query_type: QueryType = QueryType.TOTAL) -> str:
```
[the remark this answers is not in the document]
Comment (bh2smith) [reply src/fetch/period_slippage.py:51]: Except that Per TX is only ever called in unit testing.

[the remark this answers is not in the document]
Comment (bh2smith) [reply src/fetch/period_slippage.py:51]: This refactoring is only a segway into what I actually envision for this project. 

Currently the slippage query is handling both AMM slippage and the internal balance penalty. Realistically, we should be fetching the classified internal token imbalances, classifying those which are due to slippage and those which are internal in the python code. A reason for this is to be able to log the slippage penalty separately from the internal balance and illegal trading penalty.

Should really open an issue/discussion for this. 

In any case, the TOTAL is default because that is what the scripts make use of and the PER_TX is used internally for debugging and testing.

**Review by josojo — APPROVED** [review]
Looks all good to me. Thanks for the refactor
